spec/jobs/usps_auth_token_refresh_job_spec.rb

```diff
@@ -0,0 +1,81 @@
+require 'rails_helper'
+
+RSpec.describe UspsAuthTokenRefreshJob, type: :job do
+  include UspsIppHelper
+
+  let(:subject) { described_class.new }
+  let(:root_url) { 'http://my.root.url' }
+  let(:analytics) { instance_double(Analytics) }
+  let(:usps_auth_token_cache_key) { UspsInPersonProofing::Proofer::AUTH_TOKEN_CACHE_KEY }
+
+  before do
+    allow(IdentityConfig.store).to receive(:usps_ipp_root_url).and_return(root_url)
+
+    allow(Analytics).to receive(:new).
+      with(
+        user: an_instance_of(AnonymousUser),
+        request: nil,
+        session: {},
+        sp: nil,
+      ).and_return(analytics)
+  end
+
+  describe 'usps auth token refresh job' do
+    context 'when using redis as a backing store' do
+      before do |ex|
+        allow(Rails).to receive(:cache).and_return(
+          ActiveSupport::Cache::RedisCacheStore.new(url: IdentityConfig.store.redis_throttle_url),
+        )
+      end
+
+      it 'requests and sets a new token in the cache' do
+        stub_request_token
+
+        expect(analytics).to receive(
+          :idv_usps_auth_token_refresh_job_started,
+        ).once
+        expect(analytics).to receive(
+          :idv_usps_auth_token_refresh_job_completed,
+        ).once
+
+        expect(Rails.cache).to receive(:write).with(
+          usps_auth_token_cache_key,
+          an_instance_of(String),
+          hash_including(expires_in: an_instance_of(ActiveSupport::Duration)),
+        )
+
+        subject.perform
+
+        expect(WebMock).to have_requested(:post, "#{root_url}/oauth/authenticate")
+      end
+
+      it 'manually sets the expiration' do
+        allow(analytics).to receive(:idv_usps_auth_token_refresh_job_started)
+        allow(analytics).to receive(:idv_usps_auth_token_refresh_job_completed)
+
+        stub_request_token
+        subject.perform
+
+        ttl = Rails.cache.redis.ttl(usps_auth_token_cache_key)
+        expect(ttl).to be > 0
+      end
```

Comment:
I would like to go one of the following ways with this test (exclusive):

- Use real Redis & ensure it has the correct string token, as passed from the API mock, and the expected TTL
- Mock Redis, but also check that exact string token from API mock and TTL are passed in (not only for type)
- Scale back to only a unit test for the job

Reply:
I think I'm inclined to go with the first choice, just so that the test is as close to the "real world" as possible, and so that it as similar to the `proofer_spec.rb` as possible, doing similar work and following the existing pattern. I don't have a super strong feeling on that though, so if you have a preference of the other two, please let me know.

I pushed up the change to look for the token value in [d5dc587](https://github.com/18F/identity-idp/pull/8721/commits/d5dc58782ee1c64a5da5715c11bb527ea9567ae0). The expected TTL is hard to look for because it starts counting down the moment it's in effect. So choosing a number will likely make it flaky. Anything above 0 should mean there's an entry with a TTL though, if I understand correctly. When the cache doesn't have a key for it, the TTL is `-2`.

Reply:
After talking with Tomas, we think the `> 0` should cover the TTL piece.